### DISCUSSION TITLE
GoodWe GW10K-MS-30 with GM1000 Smart Meter

### DISCUSSION THREAD

**Discussion by nicve** [mletenay/home-assistant-goodwe-inverter#298] [General]
Hello. I have a GoodWe GW10K-MS-30 with a GM1000 Smart Meter wired in.
Is it possible to view the data from the GM1000 (Meter Buy and Sell values) locally without using the SEMS integration?
The latest experimental version [v0.9.9.27] successfully shows other values but no consumption data is visible.
Thanks for any advice.

**Comment by nicve** [comment]
@mletenay @piomar123 I note your work on https://github.com/marcelblijleven/goodwe
I wonder if we can communicate and work together on improving comms with GW10K-MS-30 inverters and GM1000 meters?
I have been experimenting but am a true beginner. I would like to understand how I might add/edit IDs that are reported on to perhaps scan for data that isn't currently visible.

**Reply by piomar123** [reply]
Hi, I would start with running the goodwe library in a Python script with debug logging to display all received data and detected parameters:
```python
import asyncio
import goodwe
import logging
from pprint import pprint


async def get_runtime_data():
    ip_address = '192.168.1.106'

    inverter = await goodwe.connect(ip_address, retries=60)
    print("---------------- sensors ----------------")
    pprint(inverter.sensors())
    print("---------------- settings ----------------")
    pprint(inverter.settings())
    print("---------------- read_settings_data ----------------")
    pprint(await inverter.read_settings_data())
    print("---------------- read_runtime_data ----------------")
    pprint(await inverter.read_runtime_data())

if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    asyncio.run(get_runtime_data())
```

**Reply by nicve** [reply]
Thanks @piomar123 - that is about as far as I got. 
When I query the device directly I am able to gather serial number etc that this script suggests is not being returned. 
Is there a way to query all modbus IDs or is there some sort of dump I could create for someone knowledgeable to review?

**Reply by piomar123** [reply]
Does your inverter communicate with SolarGo app? One of the ways is to sniff the communication between the inverter and the app using Wireshark or similar software and reverse engineer the exchanged data by guessing which register describes what based on the output visible in the SolarGo. The other way is to query all the registers and try to guess which data could it represent. 
I'm not aware of any simple way to query the meaning of the IDs.

**Reply by nicve** [reply]
It does communicate with SolarGo but the relevant consumption data is not visible. 
Perhaps if you can advise how to query all registers, I can match the outputs with what I can see in SEMS? As I was playing today I was able to match data with SEMS for some values but was unsure how to query outside of the registers already listed here.

**Reply by piomar123** [reply]
As far as I can see, your inverter should be handled by dt.py family. There are commands `self._READ_RUNNING_DATA` and `self._READ_METER_DATA` which request the data from the inverter in a form of bytes. Then they are parsed by `sensors` dictionary.
There is also an unit test `GW10K_MS_30_Test` that could help parsing the bytes once you have the response.

**Reply by nicve** [reply]
Thanks, yes the dt.py family is parsing Generation data beautifully. It is the consumption data from the GM1000 that I am interested in capturing. As the GM1000 is connected to the inverter via RS485, and the data is then sent to SEMS via the WiFi/LAN Kit 2.0, I assume this data must be readable via Modbus if I can Identify the register IDs. I think this might be easy to do at night time when there is no PV generation, but still live consumption data visible on SEMS. Do you have any advice if it is possible to explore all registers / dump them to a file to search/skim? Thanks so much for your replies so far.

**Reply by piomar123** [reply]
If the data you want to read is not visible in SolarGo, then it will be difficult. Probably the only way will be to query successive registers with `read_command` and trying to analyze if the returned data is valuable. If the read values change often, you could query registers multiple times and filter only those which differ over time. 
It would be nice if someone could confirm that using `read_command` cannot change any settings of the inverter just to be on the safe side.

**Reply by ttkdroid** [reply]
> Thanks, yes the dt.py family is parsing Generation data beautifully. It is the consumption data from the GM1000 that I am interested in capturing. As the GM1000 is connected to the inverter via RS485, and the data is then sent to SEMS via the WiFi/LAN Kit 2.0, I assume this data must be readable via Modbus if I can Identify the register IDs. I think this might be easy to do at night time when there is no PV generation, but still live consumption data visible on SEMS. Do you have any advice if it is possible to explore all registers / dump them to a file to search/skim? Thanks so much for your replies so far.

Take a look on this [document](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://api.library.loxone.com/downloader/file/1432/GoodWe%2520Modbus%2520Protocol%2520Customer%2520Version.pdf&ved=2ahUKEwjj0eq4zMyJAxUe4jgGHSnRDAYQFnoECBcQAQ&usg=AOvVaw2pa6IerzWPfCzXu-HkMs2k)

Probably Modbus have a specific address for the wired GM1000 because the inverter pushes the data do SEMS. A shitty alternative is to poll this data from SEMS using http calls, but the problem is that it gets the data as it is presented and not uploaded, which has a delay of 5 minutes, so if you want realtime monitoring of your load, it may not be an accurate way. 

I have the same damn problem and meter and still no proper solution. Research continues and if you found a solution let me know... will do the same.
